### PR TITLE
fix(Core/Vehicles): feat vehicle seat addon, vehicle enter/exit positions

### PR DESCRIPTION
**Pull request by sogladev**
## Changes Proposed:

This PR proposes changes to:
-  [x] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [x] Database (SAI, creatures, etc).

implement table `vehicle_seat_addon` with data from other cores, comments explain every entry with added data from acore's hardcoded fixes. This removes the hacks for some vehicle exit positions

removed correction that disallowed passengers to switch seats in Salvaged Demolisher

Change passenger enter/exit boarding so proper flags/positions are updated on other clients. see video clips in https://github.com/azerothcore/azerothcore-wotlk/issues/11159#issuecomment-2336869205

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/20029
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/14516
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/11159
- Progresses https://github.com/azerothcore/azerothcore-wotlk/issues/15831

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [x] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [x] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [x] This pull request requires further testing and may have edge cases to be tested.

vehicle exit orientation:
passenger orientation should use vehicle orientation
gnome ejected faces has vehicle orientation https://youtu.be/XXsjwyKTyF4?si=L_mQRo6uT-Ni6WEc&t=60

## How to Test the Changes:

- [x] This pull request can be tested by following the reproduction steps provided in the linked issue
- [x] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

most of this should be done with 2 characters, pref horde

horde traveler's mammoth
`.learn 61447`
note: with .gm off 
1. eject passenger, observe NPC trajectory and rp
2. board passenger, position should be updated correctly
3. eject passenger, position should be updated correctly
4. swapping seats, position should be updated correctly

vehicle ice mammoth
`.learn 59802`
save as above, exit positions are the same

love boats
note: boats take long time to reach the endpoint
.event start 8
go stormwind/darn/uc
boats should be able to eject you back on land instead of landing in the water

salvaged demolisher
.npc add 33109
be able to swap seats
load passenger, see passenger is visible in the bucket by driver
launch, see if passengers position is visible by driver

Helmsman ship quest
AC position is correct with source https://youtu.be/HZJW-htknts&t=74
`.quest add 11661`
`.go xyz 2796.8281 7021.791 5.4624508 571 1.728`
blow horn, observe passenger exit positions matches source

## Known Issues and TODO List:

- [ ] Traveler's Mammoth NPCs need improvement, missing turn animation and cooldown for RP https://github.com/azerothcore/azerothcore-wotlk/issues/15831

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by sudlud** [comment]
Please merge master to get Codestyle CI passing

**Comment by pangolp** [comment]
@sogladev If you enable other people to commit to the branch, I can help you by uploading the change, but it won't let me push due to permissions issues.

image

**Comment by sogladev** [comment]
> @sogladev If you enable other people to commit to the branch, I can help you by uploading the change, but it won't let me push due to permissions issues.
> 
> image

@pangolp  This was already enabled

**Comment by pangolp** [comment]
> > @sogladev If you enable other people to commit to the branch, I can help you by uploading the change, but it won't let me push due to permissions issues.
> > image
> 
> @pangolp This was already enabled

Strange, it didn't let me push the changes.
I got a 403 error, about permissions.
And with other branches, I had no problems.

**Comment by pangolp** [comment]
In the SQL you also had some spaces at the end of the document.

**Comment by sogladev** [comment]
looks like SQL style check failed on 7d8fb57e377e73c0e3351de59d985d3b68e04fe2 should be good now

https://github.com/azerothcore/azerothcore-wotlk/blob/784e80668d5b57640a4c7a4d6e2c7b898e4f804d/apps/ci/ci-pending.sh#L7

File: src/server/game/Entities/Unit/Unit.cpp [hunk 1]
```diff
@@ -19749,33 +19750,22 @@ void Unit::_ExitVehicle(Position const* exitPosition)
...
-            float z = GetMapHeight(x, y, pos.GetPositionZ());
-            if (z > INVALID_HEIGHT)
-            {
-                pos.Relocate(x, y, z);
-            }
+            if (seatAddon->ExitParameter == VehicleExitParameters::VehicleExitParamOffset)
+                pos.RelocateOffset({ seatAddon->ExitParameterX, seatAddon->ExitParameterY, seatAddon->ExitParameterZ, seatAddon->ExitParameterO });
+            else if (seatAddon->ExitParameter == VehicleExitParameters::VehicleExitParamDest)
+                pos.Relocate({ seatAddon->ExitParameterX, seatAddon->ExitParameterY, seatAddon->ExitParameterZ, seatAddon->ExitParameterO });
         }
```
Comment (sudlud) [inline src/server/game/Entities/Unit/Unit.cpp:19769]: you've removed the fallback if no seatAddon is available - is the DB complete for all exits now? or should we keep the fallback here for and generate an exit position in this case

**Comment by sudlud** [comment]
tested.

just some notes from my testing
- must be in the same group to ride as a passenger on the mammoth
- loveboats
```
.go c id 36817
.go c id 37964
.go c id 37981
```
-> tested with 3 characters
-> sometimes after exit one of the characters is floating-walking in the air, but only on the other character's screens watching him
-> on the first movemant of this character, everything is fine again
- salvaged demolisher
```
.npc add temp 33109
```
-> this one has no exit positions set (?) so you just exit right into the center of the vehicle

File: src/server/game/Entities/Unit/Unit.cpp (hunk 1 above)
Comment (sudlud) [reply src/server/game/Entities/Unit/Unit.cpp:19769]: otherwise we'll just exit at the current position of the player, right?

**Comment by sogladev** [comment]
If there's a seatAddon entry, it will either use a relative offset to the vehicle's position (e.g. mammoth) or absolute position (e.g. Kologarn's arm). Otherwise, it uses the vehicle's position

When entering a vehicle, a target position can be calculated based on data from dbc with the seats x,y,z `attachmentOffset`. When exiting, this is not the case. I believe the exit position is send from the server and no client data is available. So, each seat exit that's not the vehicle's position does need an entry. Also exit splines are calculated from the player's position to a calculated exit position, but the player's relative position is always (0,0,0)? I found [cmangos' commit where they use creature model data from .m2 files](https://github.com/cmangos/mangos-wotlk/commit/dbb563075cc676c8020d7eda110715b6f975b289) to calculate the local positions correctly. It might explain why exiting 
looks off and float walking. 

Does the float walking bug happen often still? should not happen every time. Triggering a movement to fix the position shouldn't be required

> salvaged demolisher exits in the middle

has no exit positions set, only the catapult enter/exit seat was not updating  the position properly

> must be in the same group to ride as a passenger on the mammoth

blizzlike and was present before PR

**Comment by sudlud** [comment]
Thanks for the feedback, to me this is a major improvement and enables us to define exit positions now if required.

The float walking bug is probably a separate issue.

In my tests everything else looked good so far :)

**Review by sudlud — APPROVED** [review]

**Pull request merged**